Standardize headers with react navigation

albullington opened the pull request:
Closes #226 by removing custom headers and replacing them with react navigation headers.

One quirky thing about this is that left aligned titles are not supported on iOS, so we might eventually decide to replace the headers on some screens like Photo Importer and Group Photos with a customized header.

jtklein commented on src/navigation/mainStackNavigation.js:
```diff
@@ -100,6 +110,9 @@ const MainStackNavigation = ( ): React.Node => (
           name="PhotoGallery"
           component={PhotoGalleryWithPermission}
+          options={{
+            headerTitle: photoGalleryHeaderTitle
```
I personally prefer when the header options are set in the screen component itself. So that the main navigation files are only about the logic of navigation. Then also if I look at a screen component I can see all of its layout including the header (e.g., including translated strings). So, if I look at the phone all of the screen I see I know is in one file (or subcomponents of course).

It was a bit easier when using class components. Then the header setting could be on top of the file, and always in the same location in the code. Now with functional components it would be in a useEffect hook in between the other hooks.

albullington replied:
I guess I don't have a strong preference on this, but I kind of like that the headers live here. Since we're trusting react navigation to render the headers in a native way, it makes sense to me that they would be set in navigation, rather than where we're rendering the screen component.

I definitely don't love the idea of putting these in `useEffect` hooks except for the ones where we need an [interaction between the header and the screen component](https://reactnavigation.org/docs/header-buttons/#header-interaction-with-its-screen-component).

kueda replied:
Definitely benefits to both approaches, but I'm inclined to err on the side of setting this config in the `Screen` definition and not the component because that seems to be what React Navigation folks do in their documentation. Let's at least try that for a while and if it's really annoying we can try something else.

kueda commented on src/components/ObsDetails/ObsDetails.js:
```diff
@@ -159,6 +158,19 @@ const ObsDetails = ( ): Node => {
+      headerRight: editIcon
+    } );
+  }, [navigation, observation, currentUser] );
+
```
Just so I understand, this is in an effect b/c `currentUser` gets fetched asynchronously? And possibly `observation` if it's remote?

albullington replied:
Not exactly... this is in an effect because that's the [recommended way](https://reactnavigation.org/docs/header-buttons/#header-interaction-with-its-screen-component) to handle interactions between a component and a react navigation header.

In this case, the component knows there is an observation with a specific uuid, and if that observation belongs to the current user we want to show the edit button in the header. The header itself doesn't have a way of knowing which observation is being displayed in the component (& whether that observation belongs to the current user), so we need to use `navigation.setOptions` here to pass the correct right header icon from the component to the header.

kueda reviewed (changes requested):
Overall I think this looks good, and woo net decrease in code. Here are some things to address:

1. StandardCamera probably should not have a header
1. Taxon detail header has no title
1. Add evidence UI is only showing the camera and microphone options, which is making it hard to test photo import

albullington commented:
Ok, a few changes here, as requested:

1. Removed header from Standard Camera
2. Taxon Detail doesn't have a header in the Figma designs, unless I'm missing something?
3. Fixed camera options modal so it's now showing all options
4. Fixed flickering in Obs Edit title when observations have not yet loaded

kueda commented:
> Taxon Detail doesn't have a header in the Figma designs, unless I'm missing something?

You're [quite right](https://www.figma.com/file/XmLaQzJ1McnsYQysB57hib/iNaturalist-Mobile-UX-Design?node-id=1205%3A28121&t=rdOIKLu0Jy7Lt4M9-4)! @abhasinat was that an oversight? That seems inconsistent with most other screens.

abhasinat commented:
No header on Obs Detail, Taxon detail, and user profile is intentional.

kueda commented:
To be clear, these views have headers, they just don't have titles in those headers. And FWIW, in you designs on Figma, you have titles in the header for Obs Detail and User Profile, so that seems inconsistent with Taxon Detail.

<img width="827" alt="Screen Shot 2022-11-21 at 2 55 26 PM" src="https://user-images.githubusercontent.com/22691/203174401-09c18ff9-0e1f-4e81-aa30-8f387bc13d36.png">

<img width="829" alt="Screen Shot 2022-11-21 at 2 55 04 PM" src="https://user-images.githubusercontent.com/22691/203174410-c5cb4d0a-8bc9-4a36-b4fc-29aafb0e6a18.png">

abhasinat commented:
Ah right, here's what the updated UI version of obs detail, user profile, and taxon detail currently look like. You can see how I updated the header to still have buttons, but no longer be sticky to the top of the screen or have a title on any of these screens. No stress on these differences for now, it may just be helpful to know the direction these screens are going in.


<img width="653" alt="image" src="https://user-images.githubusercontent.com/44380775/203399007-b19a1444-4e41-4414-9734-283a218d4550.png">
<img width="617" alt="image" src="https://user-images.githubusercontent.com/44380775/203399043-4bcb5367-14f7-400b-8555-d44923590b82.png">
<img width="517" alt="image" src="https://user-images.githubusercontent.com/44380775/203399111-d6dc2d7f-c6ab-429a-b7ee-a1ee91264837.png">
